File: main/delete-node-in-a-bst/delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    static TreeNode* deleteNode(TreeNode* root, int key) noexcept;
};

namespace {
    TreeNode*& find(TreeNode*& root, const int key) noexcept
    {
        if (!root || root->val == key) return root;

        for (auto parent = root; ; ) {
            auto& child = (key < parent->val ? parent->left : parent->right);
            if (!child || child->val == key) return child;
            parent = child;
        }
    }

    TreeNode*& findMin(TreeNode*& root) noexcept
    {
        assert(root);
        if (!root->left) return root;

        auto parent = root;
        while (parent->left->left) parent = parent->left;
        return parent->left;
    }

    TreeNode* extractMin(TreeNode*& root) noexcept
    {
        auto& pos = findMin(root);
        const auto ret = pos;
        pos = ret->right;
        ret->right = nullptr; // makes debugging easier
        return ret;
    }

    void replace(TreeNode*& oldNode, TreeNode* newNode) noexcept
    {
        newNode->left = oldNode->left;
        newNode->right = oldNode->right;
        oldNode->left = oldNode->right = nullptr; // makes debugging easier
        oldNode = newNode;
    }
}

TreeNode* Solution::deleteNode(TreeNode* root, const int key) noexcept
{
    auto& pos = find(root, key);

    if (pos) {
        if (!pos->left)
            pos = pos->right;
        else if (!pos->right)
            pos = pos->left;
        else
            replace(pos, extractMin(pos->right));
    }

    return root;
}

File: main/delete-node-in-a-bst/delete-node-in-a-bst.cs
public class Solution {
    public TreeNode DeleteNode(TreeNode root, int key)
    {
        ref var pos = ref Find(ref root, key);

        if (pos != null) {
            if (pos.left == null)
                pos = pos.right;
            else if (pos.right == null)
                pos = pos.left;
            else
                Replace(ref pos, ExtractMin(ref pos.right));
        }

        return root;
    }

    private static ref TreeNode Find(ref TreeNode root, int key)
    {
        if (root == null || root.val == key) return ref root;

        for (var parent = root; ; ) {
            ref var child = ref (key < parent.val ? ref parent.left
                                                  : ref parent.right);

            if (child == null || child.val == key) return ref child;

            parent = child;
        }
    }

    private static ref TreeNode FindMin(ref TreeNode root)
    {
        if (root.left == null) return ref root;

        var parent = root;
        while (parent.left.left != null) parent = parent.left;
        return ref parent.left;
    }

    private static TreeNode ExtractMin(ref TreeNode root)
    {
        ref var pos = ref FindMin(ref root);
        var ret = pos;
        pos = ret.right;
        ret.right = null; // makes debugging easier
        return ret;
    }

    private static void Replace(ref TreeNode oldNode, TreeNode newNode)
    {
        newNode.left = oldNode.left;
        newNode.right = oldNode.right;
        oldNode.left = oldNode.right = null; // makes debugging easier
        oldNode = newNode;
    }
}
